halo2/src/prover.rs

```diff
@@ -36,90 +35,96 @@ use std::{
 };
 
 pub use halo2_proofs::halo2curves::bn256::Bn256;
+pub use halo2_proofs::poly::commitment::Params;
 pub use halo2_proofs::poly::kzg::commitment::ParamsKZG;
-
-/// Create a halo2 proof for a given PIL, fixed column values and witness column values
-/// We use KZG ([GWC variant](https://eprint.iacr.org/2019/953)) and Keccak256
-pub struct Halo2Prover {
+pub use halo2_proofs::SerdeFormat;
+
+/// Create a halo2 proof for a given PIL, fixed column values and witness column
+/// values. We use KZG ([GWC variant](https://eprint.iacr.org/2019/953)) and
+/// Keccak256
+///
+/// This only works with Bn254, so it really shouldn't be generic over the field
+/// element, but without RFC #1210, the only alternative I found is a very ugly
+/// "unsafe" code, and unsafe code is harder to explain and maintain.
+pub struct Halo2Prover<'a, F: FieldElement> {
+    pil: &'a Analyzed<F>,
+    plaf: Plaf,
     params: ParamsKZG<Bn256>,
     vkey: Option<VerifyingKey<G1Affine>>,
 }
 
-impl Halo2Prover {
-    pub fn new(size: DegreeType) -> Self {
-        let degree = DegreeType::BITS - size.leading_zeros() + 1;
-        Self {
-            params: ParamsKZG::<Bn256>::new(degree),
-            vkey: None,
-        }
-    }
-
-    pub fn new_from_setup(input: &mut impl io::Read) -> Result<Self, io::Error> {
-        let params = ParamsKZG::<Bn256>::read(input)?;
+fn degree_bits(degree: DegreeType) -> u32 {
+    DegreeType::BITS - degree.leading_zeros() + 1
+}
 
-        Ok(Self { params, vkey: None })
-    }
+pub fn generate_setup(size: DegreeType) -> ParamsKZG<Bn256> {
+    ParamsKZG::<Bn256>::new(degree_bits(size))
+}
 
-    pub fn add_verification_key<F: FieldElement>(
-        &mut self,
-        pil: &Analyzed<F>,
+impl<'a, F: FieldElement> Halo2Prover<'a, F> {
+    pub fn new(
+        pil: &'a Analyzed<F>,
         fixed: &[(String, Vec<F>)],
-        vkey: Vec<u8>,
-    ) {
+        setup: Option<&mut dyn io::Read>,
+    ) -> Result<Self, io::Error> {
+        Self::assert_field_is_bn254();
+
+        let params = setup
+            .map(|mut setup| ParamsKZG::<Bn256>::read(&mut setup))
+            .transpose()?
+            .map(|mut params| {
+                params.downsize(degree_bits(pil.degree()));
```

Comment:
I'm also not sure it should be done here. Question: if I want to compute two separate proofs, then aggregate them, do you see that as 3 backend instances, or would I use the same? If the same, this here isn't ideal because I might want to keep a large `setup` that works for all circuits, but when making specific proofs I downsize the setup params to that specific circuit.

Reply:
If the PIL and fixed cols do change in between the separated proofs, this API is not very good. The degree is taken from the PIL, so it would require one instance per proof.

Reply:
they don't change for separate proofs, that's the whole point.

Reply:
Then I don't understand. Either PIL and degree needs to change in between proofs, or they don't. The way it is, if the setup is loaded, it is downsized. If it is generated from the factory, it is not.